blogs/treewidth.md

```diff
@@ -0,0 +1,385 @@
+
+For simplicity, in the following sections, we will use the Einstein summation formula to represent the tensor network as high dimensional arrays' multiplication:
+$$R_{i,j,k,...} = \sum_{a, b, c, ...} A_{a,...}B_{b,...}...$$
+It is easy to see that the most direct way to calculate the result is to loop over all the indices naively, which requires $O(d^N)$ operations, where $d$ is the dimension of the index and $N$ is the number of indices.
+However, such a direct calculation is not efficient.
+Considering the following simple tensor network:
+
+![](/assets/treewidth_figs/D10.png)
+
+where the tensors are represented by the circles and the indices are represented by the lines, $A, B, C, D$ are all rank-4 tensors, all indices are of dimension $D$.
+Naively loop over all the indices requires $O(D^{12})$ operations and producing no intermediate tensors.
+On the other hand, consider the following contraction order:
+
+
+![](/assets/treewidth_figs/ABCD_contraction.png)
+
+where we first contract $A$ and $B$ to get $AB$, then contract $C$ and $D$ to get $CD$, and finally contract $AB$ and $CD$ to get the result.
+In this way, the total number of operations is $O(D^{10})$, which is much smaller than the naive calculation, while the trade-off is that we need to store the intermediate tensors $AB$ and $CD$ with size of $O(D^{8})$.
+We say such a contraction is with **time complexity** of $O(D^{10})$ and **space complexity** of $O(D^{8})$, which represents the number of operations and the largest size of the intermediate tensors, respectively.
+
+In actual calculation, we prefer such binary contractions, i.e., contracting two tensors at a time, and then contract the result with another tensor, and so on, so that we can make use of BLAS libraries to speed up the calculation.
+A given contraction order can be represented as a tree, and since we only apply binary contractions, the tree is a binary tree. 
+The contraction tree can be represented as a rooted tree, where the leaves are the tensors to be contracted and the internal nodes are the intermediate tensors.
+
+### Tools for Tensor Network Contraction Order Optimization
+
+In practice, there are many tools for tensor network contraction order optimization, including:
+* [OMEinsumContractionOrder.jl](https://github.com/TensorBFS/OMEinsumContractionOrders.jl): an open-source Julia package for finding the optimal contraction order of tensor networks, is used as backend of [OMEinsum.jl](https://github.com/under-Peter/OMEinsum.jl) and [ITensorNetworks.jl](https://github.com/mtfishman/ITensorNetworks.jl).
+* [Cotengra](https://cotengra.readthedocs.io/en/latest/)[^Gray] : a python library for contracting tensor networks or einsum expressions involving large numbers of tensors.
+
+Various methods have been proposed for optimizing the contraction order, here we introduce some of them:
```

Comment:
I think the original exact exhaustive search with some performance tricks from https://journals.aps.org/pre/abstract/10.1103/PhysRevE.90.033315 is still worth mentioning, as it preceded all of these interesting more recent approaches.

Reply:
We added a brief introduction of the exhaustive search method in the revised blog and we found it quite powerful for small graphs. Would you mind us implement it in our OMEinsumContractionOrders.jl package?

Reply:
Absolutely not; it is in the public domain and you are free to copy/reimplement.